### PR TITLE
(fix): performance.now() should be a minimal op to avoid JSON deserialization overhead

### LINKED ISSUE #8548: performance.now() should be a minimal op to avoid JSON deserialization overhead
Ref https://github.com/denoland/deno/issues/8543#issuecomment-735474718

### PR DESCRIPTION
**Pull request by Soremwar**
Closes #8548

### CODE REVIEW

**Comment by Soremwar** [comment]
A million iterations

Minimal OP

`performance.now` - Time elapsed: 2335 ms
`performance.mark` - Time elapsed: 3388 ms

JSON OP

`performance.now` - Time elapsed: 5906 ms
`performance.mark` - Time elapsed: 7564 ms

File: cli/ops/timers.rs [hunk 1]
```diff
@@ -77,7 +81,7 @@ pub fn init(rt: &mut deno_core::JsRuntime) {
   super::reg_json_sync(rt, "op_global_timer_stop", op_global_timer_stop);
   super::reg_json_sync(rt, "op_global_timer_start", op_global_timer_start);
   super::reg_json_async(rt, "op_global_timer", op_global_timer);
-  super::reg_json_sync(rt, "op_now", op_now);
+  rt.register_op("op_now", metrics_op(minimal_op(op_now)));
```
Comment (bartlomieju) [inline cli/ops/timers.rs:84]: This is debatable, maybe we shouldn't take metrics of `op_now`?

File: cli/ops/timers.rs [hunk 2]
```diff
@@ -138,26 +142,36 @@ async fn op_global_timer(
...
+  state: Rc<RefCell<OpState>>,
+  // Arguments are discarded
+  _sync: bool,
+  _x: i32,
+  mut zero_copy: BufVec,
+) -> MinimalOp {
+  match zero_copy.len() {
+    0 => return MinimalOp::Sync(Err(type_error("no buffer specified"))),
+    1 => {}
+    _ => panic!("Invalid number of arguments"),
```
Comment (bartlomieju) [inline cli/ops/timers.rs:154]: Replace with error

**Review by bartlomieju — COMMENTED** [review]
@Soremwar looks good!

File: cli/ops/timers.rs (hunk 1 above)
Comment (Soremwar) [reply cli/ops/timers.rs:84]: The wonders of copy-pasta
Will fix

File: cli/ops/timers.rs (hunk 2 above)
Comment (Soremwar) [reply cli/ops/timers.rs:154]: RangeError or custom error?

File: cli/ops/timers.rs (hunk 1 above)
Comment (bartlomieju) [reply cli/ops/timers.rs:84]: @Soremwar this is ok, I'm just wondering if we add additional overhead to the `op_now`. Maybe let's keep it for now and address it if other performance problems occur

File: cli/ops/timers.rs (hunk 2 above)
Comment (bartlomieju) [reply cli/ops/timers.rs:154]: I think you could use `type_error`

**Comment by Soremwar** [comment]
Removing the extra Uint8array actually made a big difference :O

### New times: 

#### Minimal OP v1 (Extra Uint8Array)

performance.now - Time elapsed: 2335 ms
performance.mark - Time elapsed: 3388 ms

#### Minimal OP v2 (Clean)

performance.now - Time elapsed: 1677 ms
performance.mark - Time elapsed: 2921 ms

#### JSON OP

performance.now - Time elapsed: 5906 ms
performance.mark - Time elapsed: 7564 ms

**Comment by Soremwar** [comment]
Stll quite slow compared to Node, I might look inside to check what magic they are doing to achieve those result, but for all practical purposes this is solved

**Comment by bnoordhuis** [comment]
@Soremwar Node.js uses V8 fast API calls (https://github.com/nodejs/node/pull/33600). I have a proof-of-concept in https://github.com/denoland/rusty_v8/pull/511 but it's not anywhere near ready.

**Comment by kitsonk** [comment]
Yeah, so this improvement is great for now until we can support fast API.

**Review by ry — APPROVED** [review]
Nice work @Soremwar - LGTM

This is a very encouraging example of why we should support FastAPI. Ultimately it would be awesome if all of our ops could use it.

**Pull request merged**